### PR TITLE
Typing changes

### PR DESCRIPTION
**Pull request by bluetech**
Suggested changes for the pyi branch.

### CODE REVIEW

**Comment by bluetech** [comment]
I reworked `py.path.local` (still a few TODOs). I'll complete this and the rest of the files tomorrow, hopefully.

**Comment by blueyed** [comment]
Awesome!  I guess we can leave this open until you're finished, and I'll switch to your branch when working with it again.

**Comment by bluetech** [comment]
Okay, should be ready to go. Please take a look!

I ran it against pytest features. It resulted in some type errors that are fixed in this branch: https://github.com/bluetech/pytest/commits/py-typed-fixes

Once we merge this to `py` I'll send the above branch to `pytest`. To ease the transition I made sure it works both with old and new `py`. Hopefully though `pytest` could require new `py`.

File: py/_io/capture.py [hunk 1]
```diff
@@ -13,7 +13,7 @@ class TextIO(StringIO):
         def write(self, data):
             if not isinstance(data, unicode):
                 data = unicode(data, getattr(self, '_encoding', 'UTF-8'), 'replace')
-            StringIO.write(self, data)
+            return StringIO.write(self, data)
```
Comment (bluetech) [inline py/_io/capture.py:16]: I tried to avoid changes to actual code, but needed a couple anyway.

**Pull request closed without merging**